### PR TITLE
WIP: apron scala interface

### PR DESCRIPTION
**Pull request by tianhan0**
Since
- I wish to merge the Apron Interface into the latest commit of `develop` branch
- My `plugin-apron` branch was originally branched off based on a very very early commit of the `develop` branch
- My code for apron interface is an independent package, which means that it has no dependency issues on other parts of the code

I decided to start a new branch from the latest commit of `develop` branch, and move my code for apron interface there.

Essentially this is a manual `git merge`.

Since running apron requires external libraries, please refer to this for installation: https://gitlab.com/cuplv/csci7135-s17/issues/17

There is some redundant code in `ApronTest.scala`. I leave it there for further reference, in case we need more features of apron.

The point of having `AbsTerm` as parent class for `ApronLinTerm` and `ApronNonLinTerm` is to let them share a common interface such that some high-level procedures won't need to know the difference between them. Same for `AbsExpr`, `AbsCons`, `AbsDom`.

### CODE REVIEW

File: src/main/scala/edu/colorado/plv/cuanto/numdomain/apron/ApronTest.scala [hunk 1]
```diff
@@ -0,0 +1,180 @@
+package edu.colorado.plv.cuanto.numdomain.apron
```
Comment (bechang) [inline src/main/scala/edu/colorado/plv/cuanto/numdomain/apron/ApronTest.scala]: Test code should go in test/

File: src/main/scala/edu/colorado/plv/cuanto/numdomain/AbsCons.scala [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package edu.colorado.plv.cuanto.numdomain
+
+/**
+  * Created by lumber on 4/20/17.
+  */
```
Comment (bechang) [inline src/main/scala/edu/colorado/plv/cuanto/numdomain/AbsCons.scala]: Ahem, nit but you did agree to it: https://cuplv.slack.com/archives/C3S635ZD4/p1497045800481119

**Comment by octalsrc** [comment]
Based on your installation instructions, it looks like Travis needs these packages to use the Apron dependency:

- https://packages.debian.org/jessie/libgmp-dev
- https://packages.debian.org/jessie/libmpfr-dev

(the versions might need to change)

These are added to the Travis configuration in the same way z3 is added [here](https://github.com/cuplv/cuanto/blob/develop/.travis.yml#L6-L14).

**Comment by tianhan0** [comment]
TODOs:
https://git.cs.colorado.edu/cuplv/tianhans-papers-repo/issues/68

**Comment by tianhan0** [comment]
Problem: project won't compile when I pulled the [latest commit](https://github.com/cuplv/cuplv-nixpkgs/commit/eaaab12f3833d00bb02e98e853e30cc487ad401f) of cuplv-nixpkgs.

Reason: In the latest commit, there is a new script `cuplv-nixpkgs/japron/java-prefix.patch` that move `apron.jar` and `gmp.jar` into `$apron_prefix/usr/share/`.

I understand this change is intended to specify `classpath` when compiling/running the project. However, according to [this answer](https://stackoverflow.com/questions/4975379/jar-dependency-resolution-from-within-usr-share-java), `System.getProperty("java.ext.dirs")` on my Mac is not `$apron_prefix/usr/share/`.

It is actually `/Users/lumber/Library/Java/Extensions:/Library/Java/JavaVirtualMachines/jdk1.8.0_121.jdk/Contents/Home/jre/lib/ext:/Library/Java/Extensions:/Network/Library/Java/Extensions:/System/Library/Java/Extensions:/usr/lib/java`

**Comment by octalsrc** [comment]
@litho17 Yes, it seems different distributions all have different ideas of where system jars go, and `sbt` seems to completely ignore them all anyway.  My [commit](https://github.com/cuplv/cuanto/commit/0e264023b663ed8377f3752abaf4a8899d76665f) on the `system-apron` branch is intended to specifically add `/usr/share/java` to the classpath for `sbt`.  Were you testing with the `system-apron` branch?

**Comment by octalsrc** [comment]
@litho17 The `system-apron` branch of cuanto, I mean

**Comment by tianhan0** [comment]
`nix-shell '<nixpkgs>' -A cuanto`
`nix-shell -p japron`

**Comment by tianhan0** [comment]
Add contents to `CONTRIBUTING.md` on nix setup as well as manual setup for apron

**Pull request closed without merging**